src/services/mongoose/users-ip.ts

```diff
@@ -0,0 +1,60 @@
+import { User } from "@services/mongoose/schema"
+import {
+  CouldNotFindError,
+  CouldNotFindUserFromIdError,
+  PersistError,
+  RepositoryError,
+  UnknownRepositoryError,
+} from "@domain/errors"
+
+export const UsersIpRepository = (): IUsersIpRepository => {
+  const update = async (
+    userId: UserId,
+    lastConnection: Date,
+    lastIPs?: IPType[],
+  ): Promise<true | RepositoryError> => {
+    try {
+      const result = await User.updateOne(
+        { _id: userId },
+        { $set: { lastConnection, lastIPs } },
+      )
+
+      if (result.n === 0) {
+        return new CouldNotFindError("Couldn't find user")
+      }
+
+      if (result.nModified !== 1) {
+        return new PersistError("Couldn't update ip for user")
+      }
+
+      return true
+    } catch (err) {
+      return new UnknownRepositoryError(err)
+    }
+  }
+
+  const findById = async (userId: UserId): Promise<UserIp | RepositoryError> => {
+    try {
+      const result = await User.findOne({ _id: userId }, { lastIPs: 1 })
+      if (!result) {
+        return new CouldNotFindUserFromIdError(userId)
+      }
+
+      return userFromRaw(result)
+    } catch (err) {
+      return new UnknownRepositoryError(err)
+    }
+  }
+
+  return {
+    update,
+    findById,
+  }
+}
+
+const userFromRaw = (result: UserIpType): UserIp => {
+  return {
+    id: result.id as UserId,
+    lastIPs: (result.lastIPs || []) as IPType[],
```

Comment:
Is this cast really necessary?

Reply:
I'm not sure? I basically copied what was in the `mongoose/user.ts` here. still not up to speed on best practice with typing on Typescript (specially wrt our codebase)

Reply:
If a cast es necessary to make the type checker happy then this is the right place for it. Still want to avoid it as much as possible and since `IPType` is a type defined on schema I thought perhaps the compiler would be happy without it here.

Reply:
an empty array would not be a `IPType[]`